File: python/src/strands_template/intervention.py
"""Intervention Handler Implementation."""

import logging
from typing import Any

from strands.hooks.events import BeforeToolCallEvent
from strands.interventions import InterventionHandler, Proceed, Deny

logger = logging.getLogger(__name__)


class TemplateIntervention(InterventionHandler):
    """Template intervention handler implementation.

    Intervention handlers provide composable control layers for agents.
    Override lifecycle methods to intercept events and return typed
    decisions (Proceed, Deny, Guide, Confirm, Transform).

    Example:
        ```python
        from strands import Agent
        from strands_template import TemplateIntervention

        handler = TemplateIntervention()
        agent = Agent(interventions=[handler])
        ```
    """

    name = "template-intervention"

    # Override any lifecycle methods you need — not just before_tool_call.

    def before_tool_call(self, event: BeforeToolCallEvent, **kwargs: Any) -> Proceed | Deny:
        """Called before a tool is executed.

        Args:
            event: The before-tool-call event with tool_use and agent reference.
        """
        # TODO: Implement your intervention logic
        return Proceed()

File: typescript/src/intervention.ts
/**
 * Intervention Handler Implementation.
 *
 * Intervention handlers provide composable control layers for agents.
 * Override lifecycle methods to intercept events and return typed
 * decisions (proceed, deny, guide, confirm, transform).
 *
 * @example
 * ```ts
 * import { Agent } from '@strands-agents/sdk'
 * import { TemplateIntervention } from 'strands-template'
 *
 * const agent = new Agent({ interventions: [new TemplateIntervention()] })
 * ```
 */

import { InterventionHandler, InterventionActions, type BeforeToolCallEvent } from '@strands-agents/sdk'

// The SDK's root entry point doesn't export the action types directly,
// so derive them from the action creators.
type Proceed = ReturnType<typeof InterventionActions.proceed>
type Deny = ReturnType<typeof InterventionActions.deny>

export class TemplateIntervention extends InterventionHandler {
  readonly name = 'template-intervention'

  // Override any lifecycle methods you need — not just beforeToolCall.

  override beforeToolCall(_event: BeforeToolCallEvent): Proceed | Deny {
    // TODO: Implement your intervention logic.
    return InterventionActions.proceed()
  }
}
